### PR TITLE
feat(custom-js): implement PUT /processors/{processor_id}

### LINKED ISSUE #190: Implement PUT /processors/{processorID}

### PR DESCRIPTION
**Pull request by syncush**
Closes #190 

#229

### CODE REVIEW

File: src/processors/models/processorsManager.js [hunk 1]
```diff
@@ -68,6 +68,23 @@ module.exports.redownloadJSProcessor = async function(processorId) {
     }
 };
 
+module.exports.updateProcessor = async function (processorId, processor) {
+    const oldProcessor = await databaseConnector.getProcessor(processorId);
+    if (!oldProcessor) {
+        throw generateProcessorNotFoundError();
+    }
+    processor.created_at = oldProcessor.created_at;
+    if (processor.type === 'file_download') {
```
Comment (NivLipetz) [inline src/processors/models/processorsManager.js]: i wouldn't include this in this endpoint, we have the `/prcoessors/ID/download` endpoint for this

File: src/processors/models/processorsManager.js [hunk 2]
```diff
@@ -68,6 +68,23 @@ module.exports.redownloadJSProcessor = async function(processorId) {
...
+        let newJavascript = await fileManager.downloadFile(processor.file_url);
+        fileManager.validateJavascriptContent(newJavascript);
+        processor.javascript = newJavascript;
+    } else {
+        processor.file_url = undefined;
```
Comment (NivLipetz) [inline src/processors/models/processorsManager.js]: not needed

File: src/processors/models/processorsManager.js [hunk 3]
```diff
@@ -68,6 +68,23 @@ module.exports.redownloadJSProcessor = async function(processorId) {
...
+    const oldProcessor = await databaseConnector.getProcessor(processorId);
+    if (!oldProcessor) {
+        throw generateProcessorNotFoundError();
+    }
+    processor.created_at = oldProcessor.created_at;
+    if (processor.type === 'file_download') {
+        let newJavascript = await fileManager.downloadFile(processor.file_url);
+        fileManager.validateJavascriptContent(newJavascript);
+        processor.javascript = newJavascript;
+    } else {
```
Comment (NivLipetz) [inline src/processors/models/processorsManager.js]: not needed, if `processor.type !== constants.PROCESSOR_TYPE_FILE_DOWNLOAD` then file_url should be undefined

File: src/processors/models/processorsManager.js (hunk 1 above)
Comment (NivLipetz) [inline src/processors/models/processorsManager.js]: lets change to 

` if (processor.type === constants.PROCESSOR_TYPE_FILE_DOWNLOAD && processor.file_url !==  oldProcessor.file_url) {`

Comment (NivLipetz) [reply src/processors/models/processorsManager.js]: as we discussed, let's change to 

`if (processor.type === constants.PROCESSOR_TYPE_FILE_DOWNLOAD && req.query.dl) {`

File: src/processors/models/processorsManager.js (hunk 3 above)
Comment (syncush) [reply src/processors/models/processorsManager.js]: done

File: src/processors/models/processorsManager.js (hunk 1 above)
Comment (syncush) [reply src/processors/models/processorsManager.js]: done

File: tests/integration-tests/processors/processors-test.js [hunk 4]
```diff
@@ -186,6 +190,113 @@ describe('Processors api', function() {
...
+
+                const updatedFileJSContent = 'module.exports = () => 1';
+                processor.file_url = baseURL + secondFileUri;
+                nock(baseURL).get(secondFileUri).reply(200, updatedFileJSContent);
+                const updateResponse = await requestSender.updateProcessor(processorId, processor, true);
+                should(updateResponse.statusCode).equal(200);
+                should(updateResponse.body.file_url).equal(processor.file_url);
+                should(updateResponse.body.javascript).equal(updatedFileJSContent);
+            });
+            it('update a file_download type processor - change to different file_url', async function() {
```
Comment (NivLipetz) [inline tests/integration-tests/processors/processors-test.js]: seems like the same test as the one on line 215

Comment (syncush) [reply tests/integration-tests/processors/processors-test.js]: true, deleting it right now

File: tests/integration-tests/processors/processors-test.js [hunk 5]
```diff
@@ -186,6 +190,98 @@ describe('Processors api', function() {
...
+                const processorId = createResponse.body.id;
+                const jsContent = createResponse.body.javascript;
+
+                const updatedProcessor = Object.assign({}, fileDownloadProcessor);
+                updatedProcessor.name = 'new name';
+                updatedProcessor.file_url = baseURL + secondFileUri;
+                nock(baseURL).get(secondFileUri).reply(200, 'const a = 5;');
+                const updateResponse = await requestSender.updateProcessor(processorId, updatedProcessor, false);
+                should(updateResponse.statusCode).equal(200);
+                should(updateResponse.body.name).equal(updatedProcessor.name);
```
Comment (NivLipetz) [inline tests/integration-tests/processors/processors-test.js]: add another validation of `fileDownloadProcessor.description` and `fileDownloadProcessor.file_url`

File: src/processors/controllers/processorController.js [hunk 6]
```diff
@@ -54,3 +54,31 @@ module.exports.redownloadJSProcessor = async function (req, res, next) {
...
+    try {
+        dl = convertDlQueryParamToBoolean(dl);
+        const processor = await processorManager.updateProcessor(processorId, updatedProcessor, dl);
+        res.status(200).json(processor);
+    } catch (e) {
+        next(e);
+    }
+};
+
+function convertDlQueryParamToBoolean(dl) {
```
Comment (NivLipetz) [inline src/processors/controllers/processorController.js]: lets convert this logic to the openapi3 spec

File: tests/integration-tests/processors/processors-test.js (hunk 5 above)
Comment (syncush) [reply tests/integration-tests/processors/processors-test.js]: Correct me if I am wrong but I validated the `file_url` in line `#283`

File: src/processors/controllers/processorController.js [hunk 7]
```diff
@@ -54,3 +54,31 @@ module.exports.redownloadJSProcessor = async function (req, res, next) {
         next(e);
     }
 };
+
+module.exports.updateProcessor = async function (req, res, next) {
+    let { body: updatedProcessor, params: { processor_id: processorId }, query: { dl = false } } = req;
+    try {
+        dl = convertDlQueryParamToBoolean(dl);
```
Comment (NivLipetz) [inline src/processors/controllers/processorController.js]: should remove

File: src/processors/controllers/processorController.js (hunk 6 above)
Comment (syncush) [reply src/processors/controllers/processorController.js]: the `express-swagger-validator` does not convert the string to boolean, did I miss anything in the spec?

File: tests/integration-tests/processors/processors-test.js (hunk 5 above)
Comment (NivLipetz) [reply tests/integration-tests/processors/processors-test.js]: you included it perfectly, missed it

File: src/processors/controllers/processorController.js (hunk 6 above)
Comment (syncush) [reply src/processors/controllers/processorController.js]: another example can be seen in GET /processors where I have used parseInt() in order to convert the `from` and `limit` parameters to proper integers even though the spec specifies that they are integers

Comment (NivLipetz) [reply src/processors/controllers/processorController.js]: you're right

File: docs/openapi3.yaml [hunk 8]
```diff
@@ -1152,6 +1152,12 @@ paths:
         required: true
         schema:
           type: string
+      - in: query
+        name: dl
+        description: should redownload the file (relevant for file_download typed processors)
+        schema:
+          type: boolean
+          default: false
```
Comment (NivLipetz) [inline docs/openapi3.yaml]: add enum: true, false

File: src/processors/controllers/processorController.js [hunk 9]
```diff
@@ -54,3 +54,31 @@ module.exports.redownloadJSProcessor = async function (req, res, next) {
...
+        const processor = await processorManager.updateProcessor(processorId, updatedProcessor, dl);
+        res.status(200).json(processor);
+    } catch (e) {
+        next(e);
+    }
+};
+
+function convertDlQueryParamToBoolean(dl) {
+    const error = new Error(`dl query parameter expected one of ["true", "false"], but got ${dl}`);
```
Comment (NivLipetz) [inline src/processors/controllers/processorController.js]: let's move this error to openapi3 spec by adding `enum` to the query param

File: src/processors/controllers/processorController.js (hunk 6 above)
Comment (NivLipetz) [inline src/processors/controllers/processorController.js]:
```suggestion
function convertDlQueryParamToBoolean(dl) {
     return dl.toLowerCase() === 'true'
}
```

Comment (NivLipetz) [inline src/processors/controllers/processorController.js]:
```suggestion
function convertDlQueryParamToBoolean(dl) {
     return dl.toLowerCase() === 'true'
}
```

File: docs/openapi3.yaml (hunk 8 above)
Comment (syncush) [reply docs/openapi3.yaml]: do you mean I should change it to :
```
schema:
    type: string
    enum:
     - true
     - false
```

Comment (NivLipetz) [reply docs/openapi3.yaml]: na, leave it as is. The swagger-validator will throw an error if the query param included is not "true" or "false"

File: src/processors/controllers/processorController.js (hunk 9 above)
Comment (NivLipetz) [inline src/processors/controllers/processorController.js]: can remove this error (line 70 and 71). The swagger validator will throw a 400 error like this:
```
{
    "message": "Input validation error",
    "validation_errors": [
        "query/dl should be boolean"
    ]
}
```

File: tests/integration-tests/processors/processors-test.js [hunk 10]
```diff
@@ -298,14 +394,53 @@ describe('Processors api', function() {
...
+            it('bad dl query parameter value', async function() {
+                const processorId = uuid();
+                const processor = {
+                    name: 'expected to fail',
+                    description: 'bad dl value',
+                    type: PROCESSOR_TYPE_FILE_DOWNLOAD,
+                    file_url: 'http://some.url.com'
+                };
+                const updateResponse = await requestSender.updateProcessor(processorId, processor, 'bad_value');
+                should(updateResponse.statusCode).equal(400);
```
Comment (NivLipetz) [inline tests/integration-tests/processors/processors-test.js]: assert that message is

```
{
    "message": "Input validation error",
    "validation_errors": [
        "query/dl should be boolean"
    ]
}
```

File: docs/openapi3.yaml (hunk 8 above)
Comment (syncush) [reply docs/openapi3.yaml]: are you sure? I have tried your suggestion and it didn't act as expected

File: src/processors/controllers/processorController.js (hunk 6 above)
Comment (syncush) [reply src/processors/controllers/processorController.js]: done

File: src/processors/controllers/processorController.js (hunk 9 above)
Comment (syncush) [reply src/processors/controllers/processorController.js]: done

File: docs/openapi3.yaml (hunk 8 above)
Comment (syncush) [reply docs/openapi3.yaml]: done

File: tests/integration-tests/processors/processors-test.js (hunk 10 above)
Comment (syncush) [reply tests/integration-tests/processors/processors-test.js]: done

**Pull request merged**